wcomponents-core/src/main/java/com/github/bordertech/wcomponents/WFileWidget.java

```diff
@@ -110,12 +132,86 @@ protected boolean doHandleRequest(final Request request) {
 		boolean changed = value != null || current != null;
 
 		if (changed) {
-			setData(value);
-		}
+			// Reset validation fields
+			resetValidationState();
+			
+			// if fileType is supplied then validate it
+			if (hasFileTypes()) {
+				boolean validFileType = FileUtil.validateFileType(value, getFileTypes());
+				getOrCreateComponentModel().validFileType = validFileType;
+			}
+			
+			// if fileSize is supplied then validate it
+			if (hasMaxFileSize()) {
+				boolean validFileSize = FileUtil.validateFileSize(value, getMaxFileSize());
+				getOrCreateComponentModel().validFileSize = validFileSize;
+			}
+			
+			// if file is valid, the update data
+			if (isFileSizeValid() && isFileTypeValid()) {
+				setData(value);
+			} else if (current == null) {
+				// otherwise no change
+				changed = false;
+			} else {
+				changed = true;
+				setData(null);
+			}
+		} 
 
 		return changed;
 	}
 
+	/**
+	 * Reset validation state.
+	 */
+	private void resetValidationState() {
```

Comment:
Again, just retrieve the model once. Another performance improvement is to check if you have to update the model in the first place by checking if the values have changed.